### PR TITLE
implements RFC: [C++] Constructors as default initializers

### PR DESCRIPTION
**Pull request by jmgomez**

### CODE REVIEW

File: compiler/semstmts.nim [hunk 1]
```diff
@@ -2294,34 +2294,53 @@ proc semProcAux(c: PContext, n: PNode, kind: TSymKind,
...
-          localError(c.config, n.info, "constructor must return an object")
-      else:
-        typ = s.typ[1]
-      if typ.kind == tyPtr and not isCtor:
-        typ = typ[0]
-      if typ.kind != tyObject:
-        localError(c.config, n.info, pragmaName & " must be either ptr to object or object type.")
-      if typ.owner.id == s.owner.id and c.module.id == s.owner.id:
-        c.graph.memberProcsPerType.mgetOrPut(typ.itemId, @[]).add s
+  if sfCppMember * s.flags != {}:
```
Comment (Araq) [inline compiler/semstmts.nim:2344]: Time to move this into its own proc.

File: compiler/ccgstmts.nim [hunk 2]
```diff
@@ -289,14 +289,18 @@ proc potentialValueInit(p: BProc; v: PSym; value: PNode; result: var Rope) =
     #echo "New code produced for ", v.name.s, " ", p.config $ value.info
     genBracedInit(p, value, isConst = false, v.typ, result)
 
-proc genCppVarForCtor(p: BProc, v: PSym; vn, value: PNode; decl: var Rope) =
+proc genCppParamsForCtor(p: BProc; call: PNode): string = 
   var params = newRopeAppender()
```
Comment (Araq) [inline compiler/ccgstmts.nim]:
```suggestion
  result = ""
```

File: compiler/ccgstmts.nim [hunk 3]
```diff
@@ -289,14 +289,18 @@ proc potentialValueInit(p: BProc; v: PSym; value: PNode; result: var Rope) =
...
   var argsCounter = 0
-  let typ = skipTypes(value[0].typ, abstractInst)
+  let typ = skipTypes(call[0].typ, abstractInst)
   assert(typ.kind == tyProc)
-  for i in 1..<value.len:
+  for i in 1..<call.len:
     assert(typ.len == typ.n.len)
-    genOtherArg(p, value, i, typ, params, argsCounter)
+    genOtherArg(p, call, i, typ, params, argsCounter)
```
Comment (Araq) [inline compiler/ccgstmts.nim]:
```suggestion
    genOtherArg(p, call, i, typ, result, argsCounter)
```

File: compiler/ccgstmts.nim [hunk 4]
```diff
@@ -289,14 +289,18 @@ proc potentialValueInit(p: BProc; v: PSym; value: PNode; result: var Rope) =
...
+  params
```
Comment (Araq) [inline compiler/ccgstmts.nim]:
```suggestion
```

File: compiler/semstmts.nim (hunk 1 above)
Comment (jmgomez) [reply compiler/semstmts.nim:2344]: You mean the helper `sfCppMember * s.flags != {}` or the whole sem check on `constructor`, `member` and `virtual`?

Comment (Araq) [reply compiler/semstmts.nim:2344]: Everything followed by the check should be moved into its own proc.

**Pull request merged**